src/v/storage/log_manager.cc

```diff
@@ -83,12 +83,32 @@ void log_manager::trigger_housekeeping() {
     });
 }
 
+ss::future<> log_manager::clean_close(storage::log& log) {
+    auto clean_segment = co_await log.close();
+
+    if (clean_segment) {
+        vlog(
+          stlog.debug,
+          "writing clean record for: {} {}",
+          log.config().ntp(),
+          clean_segment.value());
+        co_await _kvstore.put(
+          kvstore::key_space::storage,
+          internal::clean_segment_key(log.config().ntp()),
+          serde::to_iobuf(internal::clean_segment_value{
+            .segment_name = std::filesystem::path(clean_segment.value())
+                              .filename()
```

Comment:
So the value is the filename component of the segment. Is that enough to avoid collisions?

How about if the same partition # exists with two different revisions: the revision is the dir name, not the filename and then can the same filename appear in both directories? Could it result in a spurious hit where an the earlier revision is written to the kv_store as clear, the topic is deleted and recreated, goes through an unclean shutdown but we spuriously hit on the `filename()` of the first revision and don't replay?

Reply:
Interesting case... usually this is covered by the main log removal process where we wipe out kvstore entries as well as segment on disk -- on a subsequent recreate the keys from the old instance of that ntp will be gone.

However, it's not expensive to encode the revision ID into the key as well, and it would make us more robust against any case where the removals+creations didn't happen in order, so maybe that's a worthwhile change.

Reply:
Looking more closely, we have the existing `start_offset_key` that would have the same problem if we overlapped logs with the same ntp, but I think we don't do that: controller backend is processing deletions before creations + the deletion includes a purge of kvstore values for the ntp that should always happen before the next partition starts.

Reply:
> start_offset_key that would have the same problem if we overlapped logs with the same ntp, but I think we don't do that:

scary. I mean, sounds like it is ok, but the amount of context needed to figure that out seems like a lot. I guess it's something to keep in mind for a future storage v2.